src/java/org/apache/cassandra/db/lifecycle/CompositeLifecycleTransaction.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.lifecycle;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class CompositeLifecycleTransaction
+{
+    final LifecycleTransaction mainTransaction;
+    private final AtomicInteger partsToCommitOrAbort;
+    private volatile boolean obsoleteOriginalsRequested;
+    private volatile boolean wasAborted;
+
+    public CompositeLifecycleTransaction(LifecycleTransaction mainTransaction)
+    {
+        this.mainTransaction = mainTransaction;
+        this.partsToCommitOrAbort = new AtomicInteger(0);
+        this.wasAborted = false;
+        this.obsoleteOriginalsRequested = false;
+    }
+
+    public void register(PartialLifecycleTransaction part)
+    {
+        partsToCommitOrAbort.incrementAndGet();
+    }
+
+    public void requestObsoleteOriginals()
+    {
+        obsoleteOriginalsRequested = true;
+    }
+
+    public void commitPart()
+    {
+        partCommittedOrAborted();
+    }
+
+    private void partCommittedOrAborted()
```

Comment:
what about passing here the partial transaction as a parameter ?
instead of having only a number partsToCommitOrAbort we could have a reference to all the child transactions and ensure that we don't count the same transaction twice and when we commit all of the children are in the expected state

Reply:
You can't count the same transaction twice because of the protections in `PartialLifecycleTransaction`.

There could be some value in knowing which part did not complete, but because we don't have timeouts on these things (and actually can't, as compactions can last days) there's no obvious place to surface that information.